src/box/applier.cc

```diff
@@ -1252,6 +1298,726 @@ applier_on_rollback(struct trigger *trigger, void *event)
+					     &item->decoded->raft.vclock) != 0) {
+				diag_raise();
+			}
+		} else if (type == IPROTO_OK) {
+			/* Nothing to do. */
+		} else {
+			tnt_raise(ClientError, ER_UNKNOWN_REQUEST_TYPE, type);
+		}
+	}
+	assert(data - start <= (ssize_t)msg->parse_size);
+	msg->parse_size -= data - start;
+	return tx;
+not_read:
+	/* Discard the preallocated messages. */
+	region_truncate(region, region_svp);
+	return NULL;
+}
+
+static int
+applier_decode_txs(struct applier *applier, struct applier_data_msg *msg,
+		   struct stailq *txs)
+{
+	int tx_cnt = 0;
+	while (true) {
+		struct applier_tx *tx;
+		tx = applier_parse_tx(applier, msg);
+		if (tx == NULL)
+			break;
+		stailq_add_tail_entry(txs, tx, next);
+		++tx_cnt;
+	}
+	return tx_cnt;
+}
+
+/** Applier thread reader fiber function. */
+static int
+applier_thread_reader_f(va_list ap)
+{
+	struct applier *applier = va_arg(ap, struct applier *);
+	struct applier_thread *thread = container_of(cord(), typeof(*thread),
+						     cord);
+	while (!fiber_is_cancelled()) {
+		struct applier_data_msg *msg;
+		do {
+			msg = applier_thread_next_msg(applier->tdata);
+			if (msg  == NULL) {
+				fiber_yield();
+				if (fiber_is_cancelled())
+					return 0;
+			}
+		} while (msg == NULL);
+		struct stailq txs;
+		int tx_cnt;
+		double timeout = applier->version_id < version_id(1, 7, 7) ?
+				 TIMEOUT_INFINITY :
+				 replication_disconnect_timeout();
+		try {
+			applier_read_ahead(applier, msg, timeout);
+			stailq_create(&txs);
+			tx_cnt = applier_decode_txs(applier, msg, &txs);
+		} catch (Exception *e) {
+			return -1;
+		}
+		if (tx_cnt > 0)
+			applier_thread_push_batch(thread, msg, &txs);
+		else if (tx_cnt < 0)
+			return -1;
+	}
+	return 0;
+}
+
+static void
+applier_thread_set_ready(void *data)
+{
+	struct applier_thread *thread = (struct applier_thread *)data;
+	thread->is_ready = true;
+	fiber_cond_broadcast(&thread->ready_cond);
+}
+
+/** The main applier thread fiber function. */
+static int
+applier_thread_f(va_list ap)
+{
+	struct applier_thread *thread = va_arg(ap, typeof(thread));
+	int rc = cbus_endpoint_create(&thread->endpoint, cord()->name,
+				      fiber_schedule_cb, fiber());
+	assert(rc == 0);
+
+	cbus_pair("tx_prio", cord()->name, &thread->tx_pipe,
+		  &thread->thread_pipe, applier_thread_set_ready, thread,
+		  cbus_process);
+
+	cbus_loop(&thread->endpoint);
+
+	unreachable();
+}
+
+/** Initialize and start the applier thread. */
+static void
+applier_thread_create(struct applier_thread *thread)
+{
+	static int thread_id = 0;
+	const char *name = tt_sprintf("applier_%d", ++thread_id);
+
+	memset(thread, 0, sizeof(*thread));
+
+	fiber_cond_create(&thread->ready_cond);
+
+	if (cord_costart(&thread->cord, name, applier_thread_f, thread) != 0)
+		diag_raise();
+
+	while (!thread->is_ready) {
+		if (fiber_cond_wait(&thread->ready_cond) != 0)
+			diag_raise();
+	}
+}
+
+/** Alive applier thread list. */
+static RLIST_HEAD(thread_list);
```

Comment:
No need to use a linked list. Let's switch to an array - this should make the code a little bit simpler.

Reply:
I allow up to 1000 applier threads (similar to iproto threads).
You may in theory use up a lot of threads on an anonymous replica replicating from a huge number of anonymous replicas (there's no limit on anon replica count).

I don't like the idea to have a static 1000 element array, so I'll have to dynamically allocate it.
IMO it's not that simple compared to having a linked list.

Reply:
Since the `replication_threads` option is not dynamic (can only be set at startup), you would have to allocate the array just once, which is IMO easier than dealing with lists. Anyway, lists are usually used when one needs to insert/delete elements into/from the middle of a container so using a list in this case looks redundant. Up to you.

Reply:
Ok, I agree with you after a second thought. Fixed.